src/work/Work.cpp

```diff
@@ -4,434 +4,224 @@
-                            << getUniqueName() << " successful, scheduling run";
-        scheduleRun();
-    }
-    else if (anyChildFatalFailure())
-    {
-        CLOG(DEBUG, "Work") << "some of " << mChildren.size() << " children of "
-                            << getUniqueName() << " fatally failed, scheduling "
-                            << "fatal failure";
-        scheduleFatalFailure();
-    }
-    else if (anyChildRaiseFailure())
-    {
-        CLOG(DEBUG, "Work") << "some of " << mChildren.size() << " children of "
-                            << getUniqueName() << " failed, scheduling failure";
-        scheduleFailure();
-    }
 }
 
 void
-Work::run()
+Work::clearChildren()
 {
-    if (getState() == WORK_PENDING)
-    {
-        CLOG(DEBUG, "Work") << "starting " << getUniqueName();
-        mApp.getMetrics().NewMeter({"work", "unit", "start"}, "unit").Mark();
-        onStart();
-    }
-    CLOG(DEBUG, "Work") << "running " << getUniqueName();
-    mApp.getMetrics().NewMeter({"work", "unit", "run"}, "unit").Mark();
-    setState(WORK_RUNNING);
-    onRun();
+    // FIXME upon proper implementation of WorkScheduler shutdown, consider this
+    // assert assert(allChildrenDone());
+    mChildren.clear();
+    mNextChild = mChildren.begin();
 }
 
 void
-Work::complete(CompleteResult result)
+Work::addChild(std::shared_ptr<BasicWork> child)
 {
-    CLOG(DEBUG, "Work") << "completed " << getUniqueName();
-    auto& succ =
-        mApp.getMetrics().NewMeter({"work", "unit", "success"}, "unit");
-    auto& fail =
-        mApp.getMetrics().NewMeter({"work", "unit", "failure"}, "unit");
+    bool resetIter = !hasChildren();
+    mChildren.push_back(child);
 
-    switch (result)
+    if (resetIter)
     {
-    case WORK_COMPLETE_OK:
-        setState(onSuccess());
-        break;
-    case WORK_COMPLETE_FAILURE:
-        setState(WORK_FAILURE_RETRY);
-        break;
-    case WORK_COMPLETE_FATAL:
-        setState(WORK_FAILURE_FATAL);
-        break;
-    }
-
-    switch (getState())
-    {
-    case WORK_SUCCESS:
-        succ.Mark();
-        CLOG(DEBUG, "Work")
-            << "notifying parent of successful " << getUniqueName();
-        notifyParent();
-        break;
-
-    case WORK_FAILURE_RETRY:
-        fail.Mark();
-        onFailureRetry();
-        scheduleRetry();
-        break;
-
-    case WORK_FAILURE_RAISE:
-    case WORK_FAILURE_FATAL:
-        fail.Mark();
-        onFailureRaise();
-        CLOG(DEBUG, "Work") << "notifying parent of failed " << getUniqueName();
-        notifyParent();
-        break;
-
-    case WORK_PENDING:
-        succ.Mark();
-        advance();
-        break;
-
-    case WORK_RUNNING:
-        succ.Mark();
-        scheduleRun();
-        break;
-
-    default:
-        assert(false);
-        break;
+        mNextChild = mChildren.begin();
     }
 }
 
-void
-Work::onReset()
+bool
+Work::allChildrenSuccessful() const
 {
+    return std::all_of(mChildren.begin(), mChildren.end(),
```

Comment:
The order of events is important here - parent runs and checks the state of children before they are actually removed (that happens on the next iteration of `yieldNextRunningChild`), so I think it makes sense to have these methods. Otherwise, each implementer is forced to keep track of all works it's spawning, which seems redundant.